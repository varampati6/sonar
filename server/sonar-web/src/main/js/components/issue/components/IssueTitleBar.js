/*
 * SonarQube
 * Copyright (C) 2009-2017 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
// @flow
import React from 'react';
import { Link } from 'react-router';
import IssueChangelog from './IssueChangelog';
import IssueMessage from './IssueMessage';
import SimilarIssuesFilter from './SimilarIssuesFilter';
import LocationIndex from '../../common/LocationIndex';
import Tooltip from '../../controls/Tooltip';
import { getComponentIssuesUrl } from '../../../helpers/urls';
import { formatMeasure } from '../../../helpers/measures';
import { translate, translateWithParameters } from '../../../helpers/l10n';
import type { Issue } from '../types';

type Props = {|
  issue: Issue,
  currentPopup: string,
  onFail: Error => void,
  onFilter?: (property: string, issue: Issue) => void,
  togglePopup: string => void
|};

const stopPropagation = (event: Event) => event.stopPropagation();

export default function IssueTitleBar(props: Props) {
  const { issue } = props;
  const hasSimilarIssuesFilter = props.onFilter != null;

  const locationsCount =
    issue.secondaryLocations.length +
    issue.flows.reduce((sum, locations) => sum + locations.length, 0);

  const locationsBadge = (
    <Tooltip
      overlay={translateWithParameters(
        'issue.this_issue_involves_x_code_locations',
        formatMeasure(locationsCount)
      )}
      placement="left">
      <LocationIndex>{locationsCount}</LocationIndex>
    </Tooltip>
  );

  // dirty trick :(
  const onIssuesPage = document.getElementById('issues-page') != null;

  const issueUrl = getComponentIssuesUrl(issue.project, { issues: issue.key, open: issue.key });

  return (
    <table className="issue-table">
      <tbody>
        <tr>
          <td>
            <IssueMessage
              message={issue.message}
              rule={issue.rule}
              organization={issue.organization}
            />
          </td>
          <td className="issue-table-meta-cell issue-table-meta-cell-first">
            <ul className="list-inline issue-meta-list">
              <li className="issue-meta">
                <IssueChangelog
                  creationDate={issue.creationDate}
                  isOpen={props.currentPopup === 'changelog'}
                  issue={issue}
                  togglePopup={props.togglePopup}
                  onFail={props.onFail}
                />
              </li>
              {issue.textRange != null &&
                <li className="issue-meta">
                  <span className="issue-meta-label" title={translate('line_number')}>
                    L{issue.textRange.endLine}
                  </span>
                </li>}
              {locationsCount > 0 &&
                <li className="issue-meta">
                  {onIssuesPage
                    ? locationsBadge
                    : <Link onClick={stopPropagation} target="_blank" to={issueUrl}>
                        {locationsBadge}
                      </Link>}
                </li>}
              <li className="issue-meta">
                <Link
                  className="js-issue-permalink icon-link"
                  onClick={stopPropagation}
                  target="_blank"
                  to={issueUrl}
                />
              </li>
              {hasSimilarIssuesFilter &&
                <li className="issue-meta">
                  <SimilarIssuesFilter
                    isOpen={props.currentPopup === 'similarIssues'}
                    issue={issue}
                    togglePopup={props.togglePopup}
                    onFail={props.onFail}
                    onFilter={props.onFilter}
                  />
                </li>}
            </ul>
          </td>
        </tr>
      </tbody>
    </table>
  );
}
