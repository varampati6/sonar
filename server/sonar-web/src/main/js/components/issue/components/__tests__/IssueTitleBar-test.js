/*
 * SonarQube
 * Copyright (C) 2009-2017 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import { shallow } from 'enzyme';
import React from 'react';
import IssueTitleBar from '../IssueTitleBar';

const issue = {
  line: 25,
  textRange: {
    startLine: 25,
    endLine: 26,
    startOffset: 0,
    endOffset: 15
  },
  creationDate: '2017-03-01T09:36:01+0100',
  organization: 'myorg',
  project: 'myproject',
  key: 'AVsae-CQS-9G3txfbFN2',
  rule: 'javascript:S1067',
  message: 'Reduce the number of conditional operators (4) used in the expression',
  flows: [],
  secondaryLocations: []
};

it('should render the titlebar correctly', () => {
  const element = shallow(
    <IssueTitleBar issue={issue} currentPopup="" onFail={jest.fn()} togglePopup={jest.fn()} />
  );
  expect(element).toMatchSnapshot();
});

it('should render the titlebar with the filter', () => {
  const element = shallow(
    <IssueTitleBar
      issue={issue}
      currentPopup=""
      onFail={jest.fn()}
      onFilter={jest.fn()}
      togglePopup={jest.fn()}
    />
  );
  expect(element).toMatchSnapshot();
});

it('should count all code locations', () => {
  const issueWithLocations = {
    ...issue,
    flows: [[{}, {}, {}], [{}, {}]],
    secondaryLocations: [{}, {}]
  };
  const element = shallow(<IssueTitleBar issue={issueWithLocations} />);
  expect(element.find('LocationIndex').children().text()).toBe('7');
});
