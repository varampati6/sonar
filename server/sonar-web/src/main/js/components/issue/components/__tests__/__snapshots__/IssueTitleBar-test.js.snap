// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render the titlebar correctly 1`] = `
<table
  className="issue-table"
>
  <tbody>
    <tr>
      <td>
        <IssueMessage
          message="Reduce the number of conditional operators (4) used in the expression"
          organization="myorg"
          rule="javascript:S1067"
        />
      </td>
      <td
        className="issue-table-meta-cell issue-table-meta-cell-first"
      >
        <ul
          className="list-inline issue-meta-list"
        >
          <li
            className="issue-meta"
          >
            <IssueChangelog
              creationDate="2017-03-01T09:36:01+0100"
              isOpen={false}
              issue={
                Object {
                  "creationDate": "2017-03-01T09:36:01+0100",
                  "flows": Array [],
                  "key": "AVsae-CQS-9G3txfbFN2",
                  "line": 25,
                  "message": "Reduce the number of conditional operators (4) used in the expression",
                  "organization": "myorg",
                  "project": "myproject",
                  "rule": "javascript:S1067",
                  "secondaryLocations": Array [],
                  "textRange": Object {
                    "endLine": 26,
                    "endOffset": 15,
                    "startLine": 25,
                    "startOffset": 0,
                  },
                }
              }
              onFail={[Function]}
              togglePopup={[Function]}
            />
          </li>
          <li
            className="issue-meta"
          >
            <span
              className="issue-meta-label"
              title="line_number"
            >
              L
              26
            </span>
          </li>
          <li
            className="issue-meta"
          >
            <Link
              className="js-issue-permalink icon-link"
              onClick={[Function]}
              onlyActiveOnIndex={false}
              style={Object {}}
              target="_blank"
              to={
                Object {
                  "pathname": "/project/issues",
                  "query": Object {
                    "id": "myproject",
                    "issues": "AVsae-CQS-9G3txfbFN2",
                    "open": "AVsae-CQS-9G3txfbFN2",
                  },
                }
              }
            />
          </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>
`;

exports[`should render the titlebar with the filter 1`] = `
<table
  className="issue-table"
>
  <tbody>
    <tr>
      <td>
        <IssueMessage
          message="Reduce the number of conditional operators (4) used in the expression"
          organization="myorg"
          rule="javascript:S1067"
        />
      </td>
      <td
        className="issue-table-meta-cell issue-table-meta-cell-first"
      >
        <ul
          className="list-inline issue-meta-list"
        >
          <li
            className="issue-meta"
          >
            <IssueChangelog
              creationDate="2017-03-01T09:36:01+0100"
              isOpen={false}
              issue={
                Object {
                  "creationDate": "2017-03-01T09:36:01+0100",
                  "flows": Array [],
                  "key": "AVsae-CQS-9G3txfbFN2",
                  "line": 25,
                  "message": "Reduce the number of conditional operators (4) used in the expression",
                  "organization": "myorg",
                  "project": "myproject",
                  "rule": "javascript:S1067",
                  "secondaryLocations": Array [],
                  "textRange": Object {
                    "endLine": 26,
                    "endOffset": 15,
                    "startLine": 25,
                    "startOffset": 0,
                  },
                }
              }
              onFail={[Function]}
              togglePopup={[Function]}
            />
          </li>
          <li
            className="issue-meta"
          >
            <span
              className="issue-meta-label"
              title="line_number"
            >
              L
              26
            </span>
          </li>
          <li
            className="issue-meta"
          >
            <Link
              className="js-issue-permalink icon-link"
              onClick={[Function]}
              onlyActiveOnIndex={false}
              style={Object {}}
              target="_blank"
              to={
                Object {
                  "pathname": "/project/issues",
                  "query": Object {
                    "id": "myproject",
                    "issues": "AVsae-CQS-9G3txfbFN2",
                    "open": "AVsae-CQS-9G3txfbFN2",
                  },
                }
              }
            />
          </li>
          <li
            className="issue-meta"
          >
            <SimilarIssuesFilter
              isOpen={false}
              issue={
                Object {
                  "creationDate": "2017-03-01T09:36:01+0100",
                  "flows": Array [],
                  "key": "AVsae-CQS-9G3txfbFN2",
                  "line": 25,
                  "message": "Reduce the number of conditional operators (4) used in the expression",
                  "organization": "myorg",
                  "project": "myproject",
                  "rule": "javascript:S1067",
                  "secondaryLocations": Array [],
                  "textRange": Object {
                    "endLine": 26,
                    "endOffset": 15,
                    "startLine": 25,
                    "startOffset": 0,
                  },
                }
              }
              onFail={[Function]}
              onFilter={[Function]}
              togglePopup={[Function]}
            />
          </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>
`;
